libs/utils/android/workload.py

```diff
@@ -30,113 +30,109 @@ class Workload(object):
         """
-        self.te = test_env
-        self.target = test_env.target
+        self._te = test_env
+        self._target = test_env.target
         self._log = logging.getLogger('Workload')
 
-        wloads = Workload.availables(self.target)
-        self._log.info('Workloads available on target:')
-        self._log.info('  %s', wloads)
+        # Set of data reported in output of each run
+        self.trace_file = None
+        self.nrg_report = None
+
+        self._packages = None
+        self._availables = {}
+
+        # Build list of supported workloads available in the target
+        self._check_availables()
 
     def _adb(self, cmd):
-        return 'adb -s {} {}'.format(self.target.adb_name, cmd)
+        return 'adb -s {} {}'.format(self._target.adb_name, cmd)
+
+    def _check_availables(self):
+        """
+        List the supported android workloads which are available on the target
+        """
+        # Getting the list of installed packages
+        self._packages = self._target.execute('pm list packages').splitlines()
+        self._log.debug('Packages:\n%s', self._packages)
+
+        self._log.debug('Building list of available workloads...')
+        for sc in Workload.__subclasses__():
+            self._log.debug('Checking workload [%s]...', sc.__name__)
+            if 'package:' + sc.package in self._packages:
+                self._availables[sc.__name__.lower()] = sc
+
+        self._log.info('Supported workloads available on target:')
+        self._log.info('  %s', ', '.join(self._availables.keys()))
+
+    def getInstance(self, name):
+        """
+        Get a reference to the specified Android workload
+        """
+        # Initialize list of available workloads
+        if self._packages is None:
+            self._check_availables()
+
+        if name.lower() not in self._availables:
+            self._log.warning('Workload [%s] not available on target', name)
+            return None
 
+        return self._availables[name.lower()](self._te)
 
-    def run(self, exp_dir, **kwargs):
+    def run(self, out_dir, collect='',
+            **kwargs):
         raise RuntimeError('Not implemeted')
 
+    def tracingStart(self):
+        # Start FTrace
+        if 'ftrace' in self.collect:
+            self.trace_file = os.path.join(self.out_dir, 'trace.dat')
+            self._log.info('FTrace START')
+            self._te.ftrace.start()
+        # Start Systrace (mutually exclusive with ftrace)
+        elif 'systrace' in self.collect:
+            self.trace_file = os.path.join(self.out_dir, 'trace.html')
+            # Get the systrace time
+            match = re.search(r'systrace_([0-9]+)', collect)
+            if match:
+                self._trace_time = match.group(1)
+            else:
+                # TODO: must implement a CTRL+C based systrace stopping
+                self._log.warning("Systrace time NOT defined, tracing for 10[s]")
+                self._trace_time = 10
```

Comment:
Some imports are missing for this code. AFAICT these are the imports we need for this file:

```
import os
import logging
import re

from . import System
```